src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,220 @@
+
+        private async ValueTask<string> HandleExtensionMessageInCurrentProcessAsync<TArgument>(
+            TArgument executeArgument, bool isSolution, string messageName, string jsonMessage,
+            Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>>> cachedHandlers,
+            CancellationToken cancellationToken)
+        {
+            AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>> lazyHandlers;
+            lock (_gate)
+            {
+                // May be called a lot.  So we use the non-allocating form of this lookup pattern.
+                lazyHandlers = cachedHandlers.GetOrAdd(
+                    messageName,
+                    static (messageName, arg) => AsyncLazy.Create(
+                        static (arg, cancellationToken) => arg.@this.ComputeHandlersAsync(arg.messageName, arg.isSolution, cancellationToken),
+                        (messageName, arg.@this, arg.isSolution)),
+                    (messageName, @this: this, isSolution));
+            }
+
+            var handlers = await lazyHandlers.GetValueAsync(cancellationToken).ConfigureAwait(false);
+            if (handlers.Length == 0)
+                throw new InvalidOperationException($"No handler found for message {messageName}.");
+
+            if (handlers.Length > 1)
+                throw new InvalidOperationException($"Multiple handlers found for message {messageName}.");
+
+            var handler = handlers[0];
+            if (s_disabledExtensionHandlers.TryGetValue(handler, out _))
+                throw new InvalidOperationException($"Handler was disabled due to previous exception.");
+
+            try
+            {
+                var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+                var result = await ((IExtensionMessageHandlerWrapper<TArgument>)handler).ExecuteAsync(
+                    message, executeArgument, cancellationToken).ConfigureAwait(false);
+                return JsonSerializer.Serialize(result, handler.ResponseType);
+            }
+            catch (Exception ex) when (DisableHandlerAndPropagate(ex))
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            bool DisableHandlerAndPropagate(Exception ex)
+            {
+                FatalError.ReportNonFatalError(ex, ErrorSeverity.Critical);
+
+                // Any exception thrown in this method is left to bubble up to the extension. But we unregister this handler
+                // from that assembly to minimize the impact.
+                s_disabledExtensionHandlers.TryAdd(handler, handler);
+                return false;
+            }
+        }
+
+        private async Task<ImmutableArray<IExtensionMessageHandlerWrapper>> ComputeHandlersAsync(
+            string messageName, bool isSolution, CancellationToken cancellationToken)
+        {
+            using var _1 = ArrayBuilder<ExtensionFolder>.GetInstance(out var extensionFolders);
```

Comment:
should this be a `TemporaryArray` with a set size (since we know it already)?

Reply:
it would be a challenge here as it would only be saafe to get the length under the lock, which would make this annoying to write out :)